runners/apex/src/main/java/org/apache/beam/runners/apex/ApexRunner.java

```diff
@@ -0,0 +1,166 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.apex;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import org.apache.beam.runners.apex.translators.TranslationContext;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.runners.PipelineRunner;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.transforms.windowing.WindowFn;
+import org.apache.beam.sdk.util.AssignWindows;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.hadoop.conf.Configuration;
+
+import com.datatorrent.api.Context.DAGContext;
+import com.datatorrent.api.DAG;
+import com.datatorrent.api.LocalMode;
+import com.datatorrent.api.StreamingApplication;
+import com.google.common.base.Throwables;
+
+/**
+ * A {@link PipelineRunner} that translates the
+ * pipeline to an Apex DAG and executes it on an Apex cluster.
+ * <p>
+ * Currently execution is always in embedded mode,
+ * launch on Hadoop cluster will be added in subsequent iteration.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ApexRunner extends PipelineRunner<ApexRunnerResult> {
+
+  private final ApexPipelineOptions options;
+
+  public ApexRunner(ApexPipelineOptions options) {
+    this.options = options;
+  }
+
+  public static ApexRunner fromOptions(PipelineOptions options) {
+    return new ApexRunner((ApexPipelineOptions) options);
+  }
+
+  @Override
+  public <OutputT extends POutput, InputT extends PInput> OutputT apply(
+      PTransform<InputT, OutputT> transform, InputT input) {
+    if (Window.Bound.class.equals(transform.getClass())) {
+      return (OutputT) ((PCollection) input).apply(
+          new AssignWindowsAndSetStrategy((Window.Bound) transform));
+    } else if (Create.Values.class.equals(transform.getClass())) {
```

Comment:
Ah, because of this, you can disregard my comment about this not being a primitive. But this hook in `apply` is one of those areas of the SDK that we aim to remove, because it is so interwoven with the user's process of building their pipeline. Instead, we want to build the whole thing, and then allow global search & replace.

It should be nearly the same code, just refactored, but it will make it so the difference between global replacement and translation is less, so you won't have as much motivation to do the replacements.